ngraph/core/src/op/max_unpool.cpp

```diff
@@ -0,0 +1,83 @@
+#include "ngraph/op/max_unpool.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v8::MaxUnpool, "MaxUnpool", 1);
+
+op::v8::MaxUnpool::MaxUnpool(const ngraph::Output<ngraph::Node>& poolInp,
+                             const ngraph::Output<ngraph::Node>& poolOut,
+                             const ngraph::Output<ngraph::Node>& inp,
+                             const ngraph::Output<ngraph::Node>& shape)
+    : Op({poolInp, poolOut, inp, shape}) {
+    constructor_validate_and_infer_types();
+}
+
+bool op::v8::MaxUnpool::visit_attributes(ngraph::AttributeVisitor& visitor) {
+    return true;
+}
+
+std::shared_ptr<ngraph::Node> op::v8::MaxUnpool::clone_with_new_inputs(const ngraph::OutputVector& new_args) const {
+    if (new_args.size() != 4) {
+        throw ngraph::ngraph_error("Incorrect number of new arguments");
+    }
+    return std::make_shared<MaxUnpool>(new_args.at(0), new_args.at(1), new_args.at(2), new_args.at(3));
+}
+
+void op::v8::MaxUnpool::validate_and_infer_types() {
+    auto outShape = get_input_partial_shape(3);
+    auto poolInpShape = get_input_partial_shape(0);
+    if (poolInpShape.is_static()) {  // need to fix it
+        auto inpShape = poolInpShape.to_shape();
+        outShape[0] = inpShape[0];  // Use only spatial dimensions from shape
+        outShape[1] = inpShape[1];  // and restore batch and channels
+        set_output_type(0, get_input_element_type(0), outShape);
+    }
+}
+
+bool op::v8::MaxUnpool::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const {
+    const float* poolInp = inputs[0]->get_data_ptr<float>();
```

Comment:
This is not aceotable, we should support different FP types, like fp16 for example